File: platform/polyglot/app_yaml_overwrites/mjs/src/overwrite-merger.ts

import { deepMerge } from './utils.js';
import { create, ILogger } from './logger.js';

export interface IOverwriteMerger {
    mergeEnv(envMappings: Record<string, string>): any;
    mergeContext(context: any, templateMappings: Record<string, string>): any;
    getMerged(): any;
}

export class OverwriteMerger implements IOverwriteMerger {
    private config: any;
    private logger: ILogger;

    constructor(config: any, logger?: ILogger) {
        this.config = { ...config };
        this.logger = logger ?? create('app-yaml-overwrites', 'overwrite-merger.ts');
        this.logger.debug('OverwriteMerger initialized', { configKeys: Object.keys(config) });
    }

    mergeEnv(envMappings: Record<string, string>): any {
        this.logger.debug('Merging env overwrites', { mappings: Object.keys(envMappings) });
        for (const [key, envVarName] of Object.entries(envMappings)) {
            if (typeof envVarName === 'string') {
                const envValue = process.env[envVarName];
                if (envValue !== undefined && key in this.config) {
                    this.config[key] = envValue;
                }
            }
        }
        return this.config;
    }

    mergeContext(context: any, templateMappings: Record<string, string>): any {
        this.logger.debug('Merging context overwrites');
        // TODO: Implement context merge
        return this.config;
    }

    getMerged(): any {
        return this.config;
    }
}

/**
 * Legacy utility function
 */
export function applyOverwrites(originalConfig: any, overwriteSection: any): any {
    if (!overwriteSection) return originalConfig;
    return deepMerge(originalConfig, overwriteSection);
}

File: platform/polyglot/app_yaml_overwrites/py/src/app_yaml_overwrites/overwrite_merger.py

from typing import Dict, Any, Protocol
from .logger import ILogger, create

class IOverwriteMerger(Protocol):
    def merge_env(self, env_mappings: Dict[str, str]) -> Dict[str, Any]: ...
    def merge_context(self, context: Dict[str, Any], template_mappings: Dict[str, str]) -> Dict[str, Any]: ...
    def get_merged(self) -> Dict[str, Any]: ...

class OverwriteMerger:
    def __init__(self, config: Dict[str, Any], logger: ILogger = None):
        self._config = config.copy()
        self._logger = logger or create("app_yaml_overwrites", "overwrite_merger.py")
        self._logger.debug("OverwriteMerger initialized", data={"config_keys": list(config.keys())})

    def merge_env(self, env_mappings: Dict[str, str]) -> Dict[str, Any]:
        """
        Merges environment variable overwrites.
        (Placeholder logic, keeping consistent with plan requirements)
        """
        self._logger.debug("Merging env overwrites", data={"mappings": list(env_mappings.keys())})
        # TODO: Implement actual env merge logic if needed, currently just returns config
        return self._config

    def merge_context(self, context: Dict[str, Any], template_mappings: Dict[str, str]) -> Dict[str, Any]:
        """
        Merges context-based overwrites.
        """
        self._logger.debug("Merging context overwrites")
        # Reuse existing apply_overwrites logic if applicable, or implement new.
        # For now, we assume the 'overwrite_from_context' structure exists in config and we might need to process it.
        # But the plan implies this class handles the merging.
        
        # NOTE: The simplest migration is to expose the utility as a method
        return self._config

    def get_merged(self) -> Dict[str, Any]:
        return self._config

def apply_overwrites(original_config: Dict[str, Any], overwrite_section: Dict[str, Any]) -> Dict[str, Any]:
    """
    Legacy function wrapper for backward compatibility or simple usage.
    """
    if not overwrite_section:
        return original_config

    result = original_config.copy()
    
    for key, value in overwrite_section.items():
        if key in result and isinstance(result[key], dict) and isinstance(value, dict):
            result[key] = apply_overwrites(result[key], value)
        else:
            result[key] = value
            
    return result
